File: examples/go/clean-api-go/application/commands/change_product_status.go
package commands

import (
	"context"

	bberrors "github.com/freakjazz/backbone-go/errors"
	bbex "github.com/freakjazz/backbone-go/interfaces/responses"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
	"github.com/freakjazz/clean-api-go/domain/repositories"
)

var validStatuses = map[string]bool{"active": true, "inactive": true, "discontinued": true}

type ChangeProductStatusCommand struct {
	ProductID string
	Status    string
}

type ChangeProductStatusCommandHandler struct {
	repo   repositories.IProductRepository
	logger *logging.EnhancedLogger
}

func NewChangeProductStatusCommandHandler(repo repositories.IProductRepository) *ChangeProductStatusCommandHandler {
	return &ChangeProductStatusCommandHandler{
		repo:   repo,
		logger: logging.NewEnhancedLogger("clean-api-go").WithLayer("application").WithComponent("ChangeProductStatusCommandHandler"),
	}
}

func (h *ChangeProductStatusCommandHandler) Handle(ctx context.Context, cmd ChangeProductStatusCommand) (string, *bbex.ErrorResponse) {
	log := h.logger.WithMethod("Handle")

	if !validStatuses[cmd.Status] {
		e := bbex.ErrorResponseBuilder.ValidationError("status must be one of: active, inactive, discontinued",
			bbex.ErrorOpts{Code: bberrors.IfcInvalidRequestBody.Int()})
		return "", &e
	}

	product, err := h.repo.FindByID(ctx, cmd.ProductID)
	if err != nil || product == nil {
		e := bbex.ErrorResponseBuilder.NotFound("product not found")
		return "", &e
	}

	product.Status = cmd.Status
	if _, err := h.repo.Save(ctx, product); err != nil {
		log.ErrorWithCode("Save failed", bberrors.InfraDBFailure.Int(), map[string]interface{}{"error": err.Error()})
		e := bbex.ErrorResponseBuilder.InternalServerError(err.Error())
		return "", &e
	}

	log.Info("Product status changed", map[string]interface{}{"id": product.ID, "status": cmd.Status})
	return product.ID, nil
}

File: examples/python/clean_api_python/application/commands/change_product_status.py
from dataclasses import dataclass

from backbone.errors import ErrorCodes
from backbone.application.exceptions import ValidationException, ResourceNotFoundException
from backbone.infrastructure.logging import LoggerFactory

from domain.repositories.product_repository import IProductRepository

VALID_STATUSES = {"active", "inactive", "discontinued"}


@dataclass
class ChangeProductStatusCommand:
    product_id: str
    status: str


class ChangeProductStatusCommandHandler:
    def __init__(self, repo: IProductRepository) -> None:
        self._repo = repo
        self._logger = LoggerFactory.create_for_layer(
            service_name="clean-api-python", layer="application", component="ChangeProductStatusCommandHandler",
        )

    def handle(self, cmd: ChangeProductStatusCommand) -> str:
        if cmd.status not in VALID_STATUSES:
            raise ValidationException(
                f"status must be one of: {', '.join(sorted(VALID_STATUSES))}",
                code=ErrorCodes.APP_VALIDATION_FAILURE,
            )
        product = self._repo.find_by_id(cmd.product_id)
        if not product:
            self._logger.warning("Product not found", context={"id": cmd.product_id})
            raise ResourceNotFoundException(
                "Product not found",
                resource_type="Product",
                resource_id=cmd.product_id,
            )
        product.status = cmd.status
        self._repo.save(product)
        self._logger.info("Product status changed", context={"id": product.id, "status": cmd.status})
        return product.id
